File: solitaire_java/src/main/java/Deck.java
import java.util.*;

public class Deck {
	private final CardStack cards;

	public Deck() {
		cards = new CardStack();
		for (int i = 0; i < Util.rankLookup.length; i++)
			for (int j = 0; j < Util.suitLookup.length; j++)
				cards.push(new Card(Util.rankLookup[i], Util.suitLookup[j]));
	}

	public void shuffle() {
		Random r = new Random();
		for (int i = 0; i < cards.size(); i++)
			swap(cards, i, r.nextInt(cards.size() - i) + i);
	}

	public Card draw() {
		if (cards.isEmpty())
			return null;
		return cards.pop();
	}

	public CardStack draw(int n) {
		if (n < 1 || cards.isEmpty())
			return null;
		CardStack temp = new CardStack();
		for (int i = 0; i < n && !cards.empty(); i++)
			temp.push(cards.pop());
		return temp;
	}

	private void swap(Vector l, int i1, int i2) {
		Object temp = l.get(i1);
		l.set(i1, l.get(i2));
		l.set(i2, temp);
	}
}

File: solitaire_js/src/Deck.js
class Deck {
  constructor() {
    this.cards = new CardStack();
    for (let i = 0; i < Ranks.length; i++)
      for (let j = 0; j < Suits.length; j++)
        this.cards.push(new Card(Ranks[i], Suits[j]));
  }

  shuffle() {
    for (let i = 0; i < this.cards.length; i++)
      swap(this.cards, i, getRandomInt(this.cards.length) + i);
  }

  draw() {
    return this.cards.pop();
  }

  draw(n) {
    if (n < 1 || this.cards.length == 0)
      return null;
    else {
      let temp = new CardStack();
      for (let i = 0; i < n && (this.cards.length > 0); i++)
        temp.push(cards.pop());
      return temp;
    }
  }

  swap(c, i, j) {
    let temp = c[i];
    c[i] = c[j];
    c[j] = temp;
  }
}